File: 2020/01/main.js
import { readFileSync } from "fs";

const input = readFileSync("2020/01/input.txt")
	.toString()
	.trim()
	.split("\n")
	.map((line) => parseInt(line));

const targetSum = 2020;

function part1() {
	for (let i = 0; i < input.length; i++) {
		for (let j = i + 1; j < input.length; j++) {
			if (input[i] + input[j] === targetSum) {
				return input[i] * input[j];
			}
		}
	}
}

function part2() {
	for (let i = 0; i < input.length; i++) {
		for (let j = i + 1; j < input.length; j++) {
			for (let k = j + 1; k < input.length; k++) {
				if (input[i] + input[j] + input[k] === targetSum) {
					return input[i] * input[j] * input[k];
				}
			}
		}
	}
}

console.log("part 1: " + part1());
console.log("part 2: " + part2());

File: 2020/01/main.zig
const std = @import("std");
const print = std.debug.print;

const input_file = @embedFile("input.txt");
const target_sum = 2020;

fn parseInput(allocator: *std.mem.Allocator) ![]const i32 {
    var result = std.ArrayList(i32).init(allocator);
    errdefer result.deinit();
    var lines = std.mem.tokenize(input_file, "\n");
    while (lines.next()) |line| {
        const value = try std.fmt.parseInt(i32, line, 10);
        try result.append(value);
    }
    return result.toOwnedSlice();
}

fn part1(input: []const i32) ?i32 {
    var i: usize = 0;
    while (i < input.len) {
        var j: usize = i + 1;
        while (j < input.len) {
            if (input[i] + input[j] == target_sum) {
                return input[i] * input[j];
            }
            j += 1;
        }
        i += 1;
    }
    return null;
}

fn part2(input: []const i32) ?i32 {
    var i: usize = 0;
    while (i < input.len) {
        var j: usize = i + 1;
        while (j < input.len) {
            var k: usize = j + 1;
            while (k < input.len) {
                if (input[i] + input[j] + input[k] == target_sum) {
                    return input[i] * input[j] * input[k];
                }
                k += 1;
            }
            j += 1;
        }
        i += 1;
    }
    return null;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();

    const input = try parseInput(&gpa.allocator);
    defer gpa.allocator.free(input);

    print("part 1: {}\n", .{part1(input)});
    print("part 2: {}\n", .{part2(input)});
}
